File: python/src/algorithms/url_safe_codec.py
import base64
import json
from typing import Union


def encode(data: Union[dict, str]) -> str:
    return base64.urlsafe_b64encode(json.dumps(data).encode()).decode().rstrip("=")


def decode(data: str) -> Union[dict, str]:
    needs_padding = len(data) % 4

    if needs_padding:
        padding_size = 4 - needs_padding
        data += "=" * padding_size

    return json.loads(base64.urlsafe_b64decode(data).decode())

File: javascript/src/algorithms/url_safe_codec.js
export function encode(data) {
  let base64String = Buffer.from(JSON.stringify(data), "binary").toString(
    "base64"
  );

  while (base64String.endsWith("=")) {
    base64String = base64String.slice(0, -1);
  }

  return encodeURIComponent(base64String);
}

export function decode(data) {
  let dataToDecode = decodeURIComponent(data);
  const needsPadding = dataToDecode.length % 4;

  if (needsPadding !== 0) {
    const paddingSize = 4 - needsPadding;
    dataToDecode += "=".repeat(paddingSize);
  }

  return JSON.parse(Buffer.from(dataToDecode, "base64").toString("binary"));
}
